### PR TITLE
Initial extension manager implementation

### PR DESCRIPTION
**Pull request by alanocallaghan**
Features:

- Display name, description, version
- Add extensions from a github repo[^1]
- Drag & drop repo URLs to install[^2]
- Remove and update[^3] extensions
- Open containing folder

Other changes:

- Make extensions `Map` an `Observable`
- Slightly change "no newer version" message, because 0.4.4 is not newer than 0.4.4 (for example)
- Set minimum dimensions on welcome stage

Questionable:

- New Pane class as a (static) inner class of ExtensionManager, could be separate.
- Use of FXML for `ExtensionListCellVBox` --- it's easier for me to make sensible layouts with ML, but it also seems convoluted for a small class
- Existing "Installed extensions" pane distinguishes between ImageServer providers and Extensions, this one doesn't (which I think is fine, but I could easily add a wrapper class like the old `QuPathExtensionEntry` class.
- Downloading code located in DragDropImportHandler, seems like it should be in a Utils class instead.

[^1]: If the repo follows some conventions
[^2]: ...hopefully, if the repo follows some conventions...
[^3]: if the extension class implements GitHubProject

### CODE REVIEW

**Comment by alanocallaghan** [comment]
Some unwanted commits in here too, oops!

**Comment by alanocallaghan** [comment]
Discussion in https://github.com/qupath/planner/issues/19; UI screenshot below

Screenshot from 2023-09-22 15-29-12

**Comment by alanocallaghan** [comment]
Noting to myself that I need to handle the case that somebody drags a Jar URL directly, rather than just the github repo

**Comment by alanocallaghan** [comment]
Should now handle both direct jar links and just repo links (as long as they're a `qupath-extension-[something]` link).

It does mean that people open URLs to images hosted on `https://github.com/[something]`, but I'm assuming this isn't common

Otherwise I think this is mostly g2g

**Review by petebankhead — CHANGES_REQUESTED** [review]
I've only had time to look quickly so far. It seems very promising. There are some alignment / UI issues visible on the Mac, e.g.
* 'Owner' is higher than 'Repository'
* The gap before 'Cancel' seems to be using space that 'Repository' could have
* There should be some separation between the owner text field and repository label
* After clicking 'Add extension' once, the other options appear - but then clicking it again has no effect. So it's not clear what exactly is the purpose of the button, or if it has the right text on it.
* It's confusing that there's a list, but nothing can be selected. If I click on the list, the top item can take the focus, but nothing else.
* I'm not sure if the version needs some indication of what it is, either by using a table (and column heading) or adding a `v` at the start; I think centred vertical alignment for the version and buttons would look bette
* The description text would ideally be lighter in color (e.g. using opacity)
* The purpose of the 'remove' and 'update' buttons is unclear - especially for built-in extension that can't be changed. Does update 'Check for updates' or actually update?
* I'm not sure that 'Cancel' and 'Accept' are the right terms; I assume 'Accept' basically means 'Download & install'?

image

I find the fxml a bit confusing since there isn't much visible, so I couldn't mock up a suggestion - but I think it would make sense to have 'Add extension' at the bottom - and always open (e.g. using `TitlePane` for sections). Beside the 'Owner' and 'Repository' there would be one large 'Download & Install' button, which is only enabled when an owner and repo have been specified.

In this scenario, 'Open extension directory' would be another button at the bottom - possibly spanning the full width of the dialog.

Finally, I think splitting the lists at the top into different sections would make sense, if they have different purposes. Rather than 'servers' and 'extensions', they could be split between 'Built-in extensions' and 'Optional extensions'. That might help explain the differences in behavior.

File: qupath-gui-fx/src/main/java/qupath/lib/gui/ExtensionManager.java [hunk 1]
```diff
@@ -26,52 +26,80 @@
...
 /**
  * Manage loading extensions for a QuPathGUI instance.
- * 
+ *
  * @author Pete Bankhead
- * 
+ *
  * @since v0.5.0
  */
 public class ExtensionManager {
```
[the remark this answers is not in the document]
Comment (alanocallaghan) [reply qupath-gui-fx/src/main/java/qupath/lib/gui/ExtensionManager.java:94]: Sounds good, I initially implemented it as a separate class but thought having `ExtensionManager` and `ExtensionManagerGUI` (or similar) was a bit weird. But there's not much shared functionality so can just be moved again

[the remark this answers is not in the document]
Comment (petebankhead) [reply qupath-gui-fx/src/main/java/qupath/lib/gui/ExtensionManager.java:94]: I'm not attached to `ExtensionManager` as a name for the non-UI-based one.

File: qupath-gui-fx/src/main/java/qupath/lib/gui/ExtensionManager.java [hunk 2]
```diff
@@ -246,6 +277,302 @@ public void promptToCopyFilesToExtensionsDirectory(final Collection<File> files)
...
+			}
+		}
+
+		/**
+		 * Simple class that just loads the FXML for a list cell
+		 */
+		static class ExtensionListCellVBox extends VBox {
+			private final QuPathExtension extension;
+			@FXML
+			private Button rmBtn;
```
[the remark this answers is not in the document]
Comment (alanocallaghan) [reply qupath-gui-fx/src/main/java/qupath/lib/gui/ExtensionManager.java]: Huh, I'd have thought if the list was empty, these should be missing, and that's how it is for me.

**Comment by alanocallaghan** [comment]
- [x] 'Owner' is higher than 'Repository'
- [x] The gap before 'Cancel' seems to be using space that 'Repository' could have
- [x] There should be some separation between the owner text field and repository label
- [x] After clicking 'Add extension' once, the other options appear - but then clicking it again has no effect. So it's not clear what exactly is the purpose of the button, or if it has the right text on it.
- [x] It's confusing that there's a list, but nothing can be selected. If I click on the list, the top item can take the focus, but nothing else.
- [x] I'm not sure if the version needs some indication of what it is, either by using a table (and column heading) or adding a v at the start; I think centred vertical alignment for the version and buttons would look bette
- [x] The description text would ideally be lighter in color (e.g. using opacity)
- [x] The purpose of the 'remove' and 'update' buttons is unclear - especially for built-in extension that can't be changed. 
- [x] Does update 'Check for updates' or actually update?
- [x] I'm not sure that 'Cancel' and 'Accept' are the right terms; I assume 'Accept' basically means 'Download & install'?

Screenshot from 2023-10-02 22-37-08



The titles are obviously WIP (see below). Same with the extensions dir button not being full-width (although I don't truly love that personally, it's up to you)


> Finally, I think splitting the lists at the top into different sections would make sense, if they have different purposes. Rather than 'servers' and 'extensions', they could be split between 'Built-in extensions' and 'Optional extensions'. That might help explain the differences in behavior.

I don't think "Server" vs "extension" is useful yeah, it seems a bit technical for what this GUI component is. I'd have done core vs non-core, but is there a way of distinguishing between core and non-core extensions? The only distinction I could dig out was those that implement GitHubProject (or ImageServerProvider). I could hard code them, but that seems a bit fragile. What I have gone with in the current version is "Editable" vs "non-editable", but this is... well just wrong really.

Relatedly, that means that some of the user-installed extensions also can't be updated, so those buttons will need to be disabled also.

**Comment by alanocallaghan** [comment]
Updated classifications based on classloaders

Screenshot from 2023-10-03 12-09-12

**Comment by petebankhead** [comment]
(On a call, so these comments are made with only half my attention...)

I like the main split with title panes, but think I was wrong to separate the built-in in this way - it looks odd when one list is  full and the other isn't, and sizing doesn't accommodate this.

The same list would be fine for both, as long as there is some way to clearly distinguish that some are built-in and that's why they look different.

The panes shouldn't be collapsible, to make it harder to create unexpected layouts.

I really like opening the URL on double-click, but wonder if that should be a third button? I'm apprehensive about launching a browser without the user knowing  that is what will happen.

**Comment by petebankhead** [comment]
Also, I have an invalid extension installed (the [OMERO one currently isn't compatible](https://github.com/qupath/qupath-extension-omero)). If that could be captured here to easily remove that it would be great; currently, it isn't displayed at all.

That might require more refactoring, since it is identified as invalid at the point when there is an attempt to install it... and so isn't added to the extensions list.

Finally, could / should there be a way to list additional jars that are in the extensions directory but not actual extensions? Even just a count would be informative.

I'm thinking of the scenario when a user deletes an extension, but doesn't have any knowledge of the fact that its dependencies are lingering. I know we don't have proper management of the dependencies, but some indication it's an issue would be good.

**Comment by finglis** [comment]
It looks good, got a few bugs/points (sorry ignore if any are repeats from what's already been discussed) 
1. The extension text is black in dark mode
2. For the order, should "Core Extensions" be first since this will be relevant for all users? Then filters down to those wanting to explore extensions further? I do like the idea of combining the "Manage Extension" section into a bar at the top of the "User Extensions" section.
3. Also the "Core Extensions" section could have a max height defined since we know the number of extensions?
4. Width of window on opening squishes the "Manage Extension" section, increase width or change "download" to the symbol also used in WSInfer?

**Comment by alanocallaghan** [comment]
> For the order, should "Core Extensions" be first since this will be relevant for all users? Then filters down to those wanting to explore extensions further? I do like the idea of combining the "Manage Extension" section into a bar at the top of the "User Extensions" section.

My logic was that you can't actually *do* anything with core extensions, but I remain undecided there.

The rest are all good points, unlikely to be able to address them all today

**Comment by alanocallaghan** [comment]
Thanks for the feedback - I much prefer the look of this

Remaining issues:

- Probably lots I've not noticed
- When you make the height large, it drags out into empty unused space at the bottom (pictured). Have tried various VBox.vgrow settings, but no avail.

Screenshot from 2023-10-05 16-16-21

**Comment by alanocallaghan** [comment]
One possibly VERY problematic thing about the current behaviour is the extension manager will grab the JAR regardless of what QuPath compatibility is declared in the extension. Not sure how to handle this - maybe when the extensions are refreshed, if it's not compatible, then remove it?

**Comment by petebankhead** [comment]
Looks good! I fixed the vertical sizing thing - will check out the functionality tomorrow.

**Review by petebankhead — APPROVED** [review]

**Pull request merged**